### PR TITLE
Add minimal support for Bandit web server

### PR DESCRIPTION
**Pull request by axelson**
Adds `SiteEncrypt.Phoenix.configure_https_bandit/2` to allow the ability to easily run SiteEncrypt with the Bandit web server (https://github.com/mtrudel/bandit/)

### CODE REVIEW

File: lib/site_encrypt/phoenix.ex [hunk 1]
```diff
@@ -54,6 +54,23 @@ defmodule SiteEncrypt.Phoenix do
     end
   end
 
+  @doc """
+  Merges paths to key and certificates to the `:https` configuration of the endpoint config for Bandit.
+
+  This is the same as `configure_https/2` but returns in the format that Bandit
+  (https://github.com/mtrudel/bandit/) expects.
+  """
+  defmacro configure_https_bandit(config, https_opts \\ []) do
```
Comment (mtrudel) [inline lib/site_encrypt/phoenix.ex:63]: I think you could hide the decision as to whether to configure Cowboy or Bandit from the user by looking at the value of the `:adapter` key in the passed config?

Comment (sasa1977) [reply lib/site_encrypt/phoenix.ex:63]: I agree. If this decision can be made automatically, then it should be done by the lib.

**Review by sasa1977 — COMMENTED** [review]
Thanks! A few things I find missing:

1. Optional requirement for the bandit version in mix.exs
2. A test proving that this works

File: lib/site_encrypt/phoenix.ex (hunk 1 above)
Comment (axelson) [reply lib/site_encrypt/phoenix.ex]: Ah, good idea. Added in 958d46a314620c5760fbd76e64110354ab65e32a

**Comment by axelson** [comment]
> 1. Optional requirement for the bandit version in mix.exs

Added in c5fc376e0159e820612f3c25921563ce0467a5d0

> 2. A test proving that this works

Do you have any guidance about what type of format this test should take? i.e. unit test of `configure_https/2`, a new bandit-powered phoenix application in `demos`, or something else.

File: lib/site_encrypt/phoenix.ex [hunk 2]
```diff
@@ -45,10 +45,21 @@ defmodule SiteEncrypt.Phoenix do
   """
   defmacro configure_https(config, https_opts \\ []) do
     quote bind_quoted: [config: config, https_opts: https_opts] do
+      # Default to cowboy
+      adapter = Keyword.get(config, :adapter, :cowboy)
```
Comment (mtrudel) [inline lib/site_encrypt/phoenix.ex:49]: Shouldn't the default be `Phoenix.Endpoint.Cowboy2Adapter` as at https://github.com/phoenixframework/phoenix/blob/105d62677fe7198b3277c4b7fd12b8c21d5f3905/lib/phoenix/endpoint/supervisor.ex#L127?

File: lib/site_encrypt/phoenix.ex [hunk 3]
```diff
@@ -45,10 +45,21 @@ defmodule SiteEncrypt.Phoenix do
...
+
       https_config =
-        (Keyword.get(config, :https) || [])
-        |> Config.Reader.merge(https_opts)
-        |> Config.Reader.merge(SiteEncrypt.https_keys(__MODULE__))
+        case adapter do
+          :cowboy ->
```
Comment (mtrudel) [inline lib/site_encrypt/phoenix.ex:53]: ibid.

File: lib/site_encrypt/phoenix.ex [hunk 4]
```diff
@@ -45,27 +45,21 @@ defmodule SiteEncrypt.Phoenix do
...
+            (Keyword.get(config, :https) || [])
+            |> Config.Reader.merge(https_opts)
+            |> Config.Reader.merge(SiteEncrypt.https_keys(__MODULE__))
+
+          Bandit.PhoenixAdapter ->
+            (Keyword.get(config, :https) || [])
+            |> Config.Reader.merge(https_opts)
+            |> Config.Reader.merge(transport_options: SiteEncrypt.https_keys(__MODULE__))
```
[the remark this answers is not in the document]
Comment (mtrudel) [reply lib/site_encrypt/phoenix.ex:61]: There are some other differences but in this context, yes. That's the only difference.

**Comment by sasa1977** [comment]
> Do you have any guidance about what type of format this test should take? i.e. unit test of `configure_https/2`, a new bandit-powered phoenix application in `demos`, or something else.

I'll try to look into this, but can't commit on time.

**Comment by sasa1977** [comment]
This took me down the rabbit hole, because after 2 years of not touching anything, the CI broke because of various issues. I had to upgrade the base stack (Erlang/Elixir/deps) on master, and then fix some issues. I merged the new master back to this branch and expanded the tests to include bandit. These tests start the actual endpoint, so I think this proves that the stuff is working correctly. Regardless, could you please verify that a bandit powered site actually works now?

**Comment by axelson** [comment]
@sasa1977 Thank you! CI failures are definitely an annoying rabbit hole!

I am pleased to confirm that the latest commit from this branch is successfully running in "prod" (i.e. my hobby sites):

https://github.com/axelson/vps/pull/2

File: lib/site_encrypt/phoenix.ex (hunk 2 above)
Comment (axelson) [reply lib/site_encrypt/phoenix.ex]: Ah, I didn't see that :+1:

**Comment by axelson** [comment]
@sasa1977 is there anything that is blocking this from being merged now? Phoenix 1.7 and MainProxy 0.3.0 are now released and it would be nice to be able to use SiteEncrypt with Bandit.

**Comment by sasa1977** [comment]
Sorry for not merging sooner. It was a busy time and I completely forgotten about this :( 
It's been awhile since I've last released, so I need to go through the commits and assemble a changelog. I'll try to publish to hex during the weekend.

Note that Phoenix 1.6 and 1.7 are not officially supported. If they are working it's by chance :-) But to be certain that everything works, we'd need to upgrade the deps and ensure that all the tests are passing. This isn't required for the immediate hex publish, but it's something that should be done going forward. Pull requests are welcome :-)

**Comment by axelson** [comment]
No worries! I've waited significantly longer to merge PRs on some of the projects I help maintain :smile:

> Note that Phoenix 1.6 and 1.7 are not officially supported. If they are working it's by chance :-) But to be certain that everything works, we'd need to upgrade the deps and ensure that all the tests are passing.

I've created https://github.com/sasa1977/site_encrypt/pull/46 to update the deps and all the tests pass, although I didn't run the docker tests but they appear to have succeeded on the PR at least.

**Comment by sasa1977** [comment]
FYI I've pushed 0.5.0 to hex

**Pull request merged**